### PR TITLE
Increase Archive time period to 15 months and Deep Archive to 3 years

### PR DESCRIPTION
**Pull request by bethanyaconnor**
This PR is a revival of https://github.com/code-dot-org/code-dot-org/pull/51013 (https://github.com/code-dot-org/code-dot-org/commit/ec671512f846a9837286ec4e83a1c6d86920ca8d#diff-36e7a5950fff79c82d69deec73fb7be176df02a96173bc67bfd5f23cec94b44fL14 is the relevant commit). I checked in S3 and the current "Days until transition to Archive Access tier" is 475 and the "Days until transition to Deep Archive Access tier" is 730 (both matching what is in this PR but not what is in staging). I adjusts the latter to match production, instead of Suresh's original commit, but happy to revert that.

tl;dr I _think_ this brings this cloudformation template in line with what is currently configured on S3.

### Other background

[Slack thread for original issue](https://codedotorg.slack.com/archives/C0T0PNTM3/p1679605311308529)

[Slack thread for reviving the PR](https://codedotorg.slack.com/archives/C03CK49G9/p1708966831159949)

[Jira ticket](https://codedotorg.atlassian.net/browse/ACQ-1642)

### CODE REVIEW

**Review by Hamms — APPROVED** [review]
I see the same numbers in S3 that you saw, so I'd say I also _think_ that this brings the template in line with what we actually have.

image

It would ideally be nice to get confirmation from @cat5inthecradle that we did indeed successfully deploy the changes from the last PR but just didn't get it merged in time; I don't think that's blocking, however.

Thanks for cleaning up some of our loose threads!

**Comment by sureshc** [comment]
My memory is that we tried to apply updates to this template and encountered issues that prevented the CloudFormation template from updating successfully due to drift between the actually provisioned resources and what was specified in the template or some other technical debt.

**Comment by cat5inthecradle** [comment]
Regarding deployment, this template doesn't appear to be documented anywhere. It references two .rb files, which would need to be packaged with the template to deploy.

I propose we move these three files to "aws/cloudformation/standalone/access_logs" where we can also include a readme or bash script to explain how to run the `cloudformation package` and `cloudformation deploy`.  ([See the PR where this was added](https://github.com/code-dot-org/code-dot-org/pull/41119))

**Comment by cat5inthecradle** [comment]
Running drift detection on the AccessLogs stack reveals what we expect. Only this setting seems to have drifted:
image

**Review by cat5inthecradle — APPROVED** [review]
Approving this, however, we should update our documentation with the correct deploy procedure.

The way we typically handle changes like this, that don't have automatic deployment, is to manually deploy the change, and only merge once it's successful. That way if we need to iterate or fix anything, we get it all into one PR.

File: aws/cloudformation/standalone/access_logs/access_logs.yml [hunk 1]
```diff
@@ -132,7 +132,7 @@ Resources:
           DATABASE: !Ref AccessLogDatabase
           TABLE: !Ref AccessLogTable
           PREFIX: !Ref BucketPrefix
-      Runtime: ruby2.7
+      Runtime: ruby3.3
```
Comment (cat5inthecradle) [inline aws/cloudformation/standalone/access_logs/access_logs.yml:136]: We cannot deploy new Ruby 2.7 lambdas, so we need to update the runtime. I have not actually tested to see if the code is valid Ruby 3.3.

File: lib/cdo/aws/cloud_formation.rb [hunk 2]
```diff
@@ -23,7 +23,7 @@ class CloudFormation
     # @return [Hash]
     attr_reader :options
 
-    # Default S3 bucket for uploading temporary template files.
+    # Default S3 bucket for uploading template files.
     TEMP_S3_BUCKET = 'cf-templates-p9nfb0gyyrpf-us-east-1'
```
Comment (cat5inthecradle) [inline lib/cdo/aws/cloud_formation.rb:27]: This is actually the permanent location of uploaded lambda code, it's inaccurate to refer to this as temporary storage. Deleting the contents would break things.

**Comment by cat5inthecradle** [comment]
@bethanyaconnor Apologies for pushing to your branch, but I went ahead and added a deploy script and made the necessary Ruby runtime version upgrade.

I think we need to test somehow that these ruby scripts are valid with Ruby 3.3. I'm not totally sure the best way to do that. We really don't want them to break, but unfortunately they must be updated if we want to run a stack update here.

**Comment by bethanyaconnor** [comment]
no worries on pushing to this branch -- this PR is mainly to tie up a 1.5 year loose end that ended up in our backlog 😅 Is this PR good to merge once I get a green Drone run? (the run failed due to the sauce connect upgrade, so hopefully just need to merge in staging)

**Comment by cat5inthecradle** [comment]
We'll want to deploy before we merge this. I think there's a pretty good chance there's something we overlooked that will fail during the deploy, and it's nice to keep it all wrapped up in one PR.

Once the drone run is green, we can give it a try.

**Comment by bethanyaconnor** [comment]
Drone is green!

**Comment by cat5inthecradle** [comment]
Some improvements were made to this lambda stack last week, I've resolved those conflicts. I think we're ready to just give this a try.

We should deploy it manually per the local [README](https://github.com/code-dot-org/code-dot-org/blob/91e79aba0ab251959b241127f937e7ec232caabf/aws/cloudformation/standalone/access_logs/README.md) before merging. Resolve problems in this PR before merging.

**Comment by cat5inthecradle** [comment]
Actually, we did this last week - the only effective change in this PR now is the ruby version upgrade - we can close this and do the upgrade in it's own PR.
